pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go

```diff
@@ -329,35 +349,94 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
 	return libsToInject
 }
 
-// extractLibInfo returns the language, the image,
-// and a boolean indicating whether the library should be injected into the pod
-func (w *Webhook) extractLibInfo(pod *corev1.Pod) ([]libInfo, bool) {
+type libInfoSource int
+
+const (
+	libInfoSourceNone libInfoSource = iota
+	libInfoSourceLibInjection
+	libInfoSourceSingleStepInstrumentation
+	libInfoSourceSingleStepLangaugeDetection
+)
+
+func (s libInfoSource) injectionType() string {
+	switch s {
+	case libInfoSourceSingleStepInstrumentation, libInfoSourceSingleStepLangaugeDetection:
+		return singleStepInstrumentationInstallType
+	case libInfoSourceLibInjection:
+		return localLibraryInstrumentationInstallType
+	default:
+		return "unknown"
+	}
+}
+
+func (s libInfoSource) autodetected() bool {
+	return s == libInfoSourceSingleStepLangaugeDetection
+}
+
+func (s libInfoSource) mutatePod(pod *corev1.Pod) error {
+	_ = mutatecommon.InjectEnv(pod, corev1.EnvVar{
+		Name:  instrumentationInstallTypeEnvVarName,
+		Value: s.injectionType(),
+	})
+	return nil
+}
+
+type extractedPodLibInfo struct {
+	// libs are the libraries we are going to attempt to inject into the given pod.
+	libs []libInfo
+	// languageDetection is set when we ran/used the language-detection annotation.
+	languageDetection *libInfoLanguageDetection
+	// source is where we got the data from, used for telemetry later.
+	source libInfoSource
+}
+
+func (e extractedPodLibInfo) withLibs(l []libInfo) extractedPodLibInfo {
+	e.libs = l
+	return e
+}
+
+// extractLibInfo metadata about what library information we should be
+// injecting into the pod and where it came from.
+func (w *Webhook) extractLibInfo(pod *corev1.Pod) extractedPodLibInfo {
+	var (
+		ssiEnabled        = w.isEnabledInNamespace(pod.Namespace)
+		langaugeDetection = w.getLibrariesLanguageDetection(pod)
+		extracted         = extractedPodLibInfo{
+			source:            libInfoSourceLibInjection,
+			languageDetection: &langaugeDetection,
+		}
+	)
+
+	if ssiEnabled {
+		extracted.source = libInfoSourceSingleStepInstrumentation
+	}
+
 	// If the pod is "injectable" and annotated with libraries to inject, use those.
 	if w.isPodEligible(pod) {
```

Comment:
Is there a case where`w.isPodEligible(pod)` can return `false` ? 

I thought we were dropping the mutation request and returning directly before reaching the call to extractLibInfo (see [here](https://github.com/DataDog/datadog-agent/blob/6aad87436465e0ee064c9bd860a6507f54a2680f/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L175-L177))

Reply:
`isPodEligible` is a superset of ssiEnabled. It is really checking to see if we should respect the annotations on the pod.

Reply:
Yes but if `isPodEligible(pod)` returns `false`, we will never reach this part of the code because a similar check was made [here](https://github.com/DataDog/datadog-agent/blob/6aad87436465e0ee064c9bd860a6507f54a2680f/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L175-L177).

In other words, my point is that maybe checking `w.isPodEligible(pod)` here is redundant because we are sure it is true

Reply:
I see! Yeah, the contract of `extractLibInfo` is now a little different, it was doing a second eligibility check even though it didn't need to at all.